sway-core/src/ir_generation/function.rs

```diff
@@ -1354,6 +1480,458 @@ impl<'eng> FnCompiler<'eng> {
+                                AsmInstruction {
+                                    op_name: Ident::new_no_span("sw".into()),
+                                    args: vec![addr_ident.clone(), item_len_ident.clone()],
+                                    immediate: Some(Ident::new_no_span("i0".into())),
+                                    metadata: None,
+                                },
+                                // advance addr
+                                AsmInstruction {
+                                    op_name: Ident::new_no_span("addi".into()),
+                                    args: vec![addr_ident.clone(), addr_ident.clone()],
+                                    immediate: Some(Ident::new_no_span("i8".into())),
+                                    metadata: None,
+                                },
+                                // load data ptr
+                                AsmInstruction {
+                                    op_name: Ident::new_no_span("lw".into()),
+                                    args: vec![data_ptr_ident.clone(), item_ptr_ident.clone()],
+                                    immediate: Some(Ident::new_no_span("i0".into())),
+                                    metadata: None,
+                                },
+                                // mcp data
+                                AsmInstruction {
+                                    op_name: Ident::new_no_span("mcp".into()),
+                                    args: vec![addr_ident, data_ptr_ident, item_len_ident.clone()],
+                                    immediate: None,
+                                    metadata: None,
+                                },
+                                // increase len
+                                AsmInstruction {
+                                    op_name: Ident::new_no_span("addi".into()),
+                                    args: vec![new_len_ident.clone(), len_ident],
+                                    immediate: Some(Ident::new_no_span("i8".into())),
+                                    metadata: None,
+                                },
+                                AsmInstruction {
+                                    op_name: Ident::new_no_span("add".into()),
+                                    args: vec![
+                                        new_len_ident.clone(),
+                                        new_len_ident.clone(),
+                                        item_len_ident,
+                                    ],
+                                    immediate: None,
+                                    metadata: None,
+                                },
+                            ],
+                            uint64,
+                            Some(new_len_ident),
+                        )
+                    }
+                    _ => return Err(CompileError::EncodingUnsupportedType { span }),
```

Comment:
The `span` currently points to the whole intrinsic call. This makes the overall message very confusing because it is not clear that the passed data is of a non-supported type. E.g.:
```
let buffer = __encode_buffer_append(buffer, my_struct);
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ Type not supported with __encode_buffer_append.
```
Just pointing to the argument would already be an improvement.

My suggestion is here to use expressive diagnostic and explain to programmers the issue and what is actually supported and expected. This is just a few minutes of extra programming effort and we will gain a lot of clarity in case of a wrong usage.

As example for a similar case, the errors we are giving for unsupported types in `match` expressions can be taken (E.g., This value has type "MyStruct". Appending values of this type to encoding buffer is not supported.). The help can list the supported types, similar to `MatchedValueIsNotValid` error.

IMO, same as `AbiEncode` implementations we have now, programmers will be writing their own impls looking at the examples in the standard library. I expect clarifying questions coming similar to those we had with `AbiEncode` before, and some can surely be avoided if the provided error messages are clear and instructive enough.

Reply:
Span: done.

Reply:
> IMO, same as AbiEncode implementations we have now, programmers will be writing their own impls looking at the examples in the standard library.

In theory, no one else needs to use this intrinsic. As nobody going to have access to the "buffer" field. More likely they will do "1u8.abi_encode(buffer)" like I do for Vecs.

Reply:
In theory :smile: I agree with the intended usage, which is always `<value>.abi_encode(buffer)` and no need for intrinsic. But I have that unsettling feeling of the habit we all devs have, to first look at `codec.sw` and then eventually read the docs :smile: That's why I am always for a good guiding error message, like the new one is.